components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/entity/ApplicationPkcs7SignatureEntity.java

```diff
@@ -0,0 +1,125 @@
+import org.bouncycastle.cms.CMSProcessableByteArray;
+import org.bouncycastle.cms.CMSSignedData;
+import org.bouncycastle.cms.CMSSignedDataGenerator;
+import org.bouncycastle.cms.CMSTypedData;
+
+public class ApplicationPkcs7SignatureEntity extends MimeEntity {
+    
+    private static final String CONTENT_DISPOSITION = "attachment; filename=\"smime.p7s\"";
+    
+    private static final String CONTENT_DESCRIPTION = "S/MIME Cryptographic Signature";
+    
+    private byte[] signature;
+    
+    public ApplicationPkcs7SignatureEntity(MimeEntity data, CMSSignedDataGenerator signer, String charset, String contentTransferEncoding, boolean isMainBody) throws HttpException {
+        Args.notNull(data, "Data");
+        Args.notNull(signer, "Signer");
+        
+        ContentType contentType = ContentType.parse(EntityUtils.appendParameter(AS2MediaType.APPLICATION_PKCS7_SIGNATURE, "charset",  charset));
+        setContentType(contentType.toString());
+        setContentTransferEncoding(contentTransferEncoding);
+        addHeader(AS2Header.CONTENT_DISPOSITION, CONTENT_DISPOSITION);
+        addHeader(AS2Header.CONTENT_DESCRIPTION, CONTENT_DESCRIPTION);
+        setMainBody(isMainBody);
+        try {
+            this.signature = createSignature(data, signer);
+        } catch (Exception e) {
+            throw new HttpException("Failed to create signed data", e);
+        }
+    }
+    
+    public ApplicationPkcs7SignatureEntity(String charset,
+                                           String contentTransferEncoding,
+                                           byte[] signature,
+                                           boolean isMainBody)
+            throws HttpException {
+        this.signature = signature;
+        ContentType contentType = ContentType
+                .parse(EntityUtils.appendParameter(AS2MediaType.APPLICATION_PKCS7_SIGNATURE, "charset", charset));
+        setContentType(contentType.toString());
+        setContentTransferEncoding(contentTransferEncoding);
+        addHeader(AS2Header.CONTENT_DISPOSITION, CONTENT_DISPOSITION);
+        addHeader(AS2Header.CONTENT_DESCRIPTION, CONTENT_DESCRIPTION);
+        setMainBody(isMainBody);
+    }
+    
+    public byte[] getSignature() {
+        return signature;
+    }
+
+    @Override
+    public void writeTo(OutputStream outstream) throws IOException {
+        NoCloseOutputStream ncos = new NoCloseOutputStream(outstream);
+
+        // Write out mime part headers if this is not the main body of message.
+        if (!isMainBody()) {
+            try (CanonicalOutputStream canonicalOutstream = new CanonicalOutputStream(ncos, AS2Charset.US_ASCII)) {
+
+                HeaderIterator it = headerIterator();
+                while (it.hasNext()) {
+                    Header header = it.nextHeader();
+                    canonicalOutstream.writeln(header.toString());
+                }
+                canonicalOutstream.writeln(); // ensure empty line between
```

Comment:
```
NoCloseOutputStream ncos = new NoCloseOutputStream(outstream);
+        try (CanonicalOutputStream canonicalOutstream = new CanonicalOutputStream(ncos, AS2Charset.US_ASCII)) {
+
+            // Write out mime part headers if this is not the main body of
+            // message.
+            if (!isMainBody()) {
+                HeaderIterator it = headerIterator();
+                while (it.hasNext()) {
+                    Header header = it.nextHeader();
+                    canonicalOutstream.writeln(header.toString());
+                }
+                canonicalOutstream.writeln(); // ensure empty line between
+                                              // headers and body; RFC2046 -
+                                              // 5.1.1
```

seems to be duplicated 5 times, maybe worth refactoring it

Reply:
Perhaps. We'll see going forward.